relay-server/src/utils/buffer.rs

```diff
@@ -28,20 +30,25 @@ impl std::error::Error for BufferError {}
 /// into the processing pipeline, use [`BufferGuard::enter`].
 #[derive(Debug)]
 pub struct BufferGuard {
-    inner: Semaphore,
+    inner: Arc<tokio::sync::Semaphore>,
```

Comment:
As mentioned above, I would revert back to the old Semaphore. Smaller PR and no more "blocking" `enter()`.